### PR TITLE
refactor(config): eliminate hardcoded cache size

### PR DESCRIPTION
**Pull request by lmatz**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
#5249 wants to eliminate hardcoded values.

This PR moves
`HASH_AGG_CACHE_SIZE`,
`JOIN_CACHE_CAP`,
`EXTREME_CACHE_SIZE`,
to the configuration file.

Makes #5080  easier to enable the no-cache test.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* RisingWave cluster configuration changes

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.
`HASH_AGG_CACHE_SIZE`, `JOIN_CACHE_CAP`, and `EXTREME_CACHE_SIZE` can be specified via `src/config/risingwave.toml` in the `[streaming.developer]` subsection.

## Refer to a related PR or issue link (optional)
#5249

### CODE REVIEW

**Comment by BugenZhao** [comment]
IIUC, with the new memory management mechanism, we'll not have `HASH_AGG_CACHE_SIZE` or `JOIN_CACHE_CAP` to configure. cc @yuhao-su  🤔

**Comment by lmatz** [comment]
I suggest we preserve a way, which could be other approaches, to enforce the cache size for a period in the near future:
1. to understand how much better the new dynamic way of adjusting cache size is
2. to get the performance numbers under certain configurations, for demonstration purposes, in case the new way is not behaving as expected

**Comment by fuyufjh** [comment]
> I suggest we preserve a way, which could be other approaches, to enforce the cache size for a period in the near future:
> 
> 1. to understand how much better the new dynamic way of adjusting cache size is
> 2. to get the performance numbers under certain configurations, for demonstration purposes, in case the new way is not behaving as expected
> 
> When it gets stabilized, we remove the cache size configuration.

These `SIZE`s are defined with the number of items instead of memory usage i.e. bytes. If you hope to expose them to users, I would like to refactor them to use bytes, otherwise, it seems pretty tricky both for us and users. For example, now we have limited the size of hash map in HashJoin to 65536 but the memory can explode because the rows inside of a join entry was not limited.

**Comment by fuyufjh** [comment]
(cont.) https://github.com/risingwavelabs/risingwave/pull/5057 may be a reference for that. It estimated the memory usage in bytes but didn't use it as the weight of LRU cache

**Comment by yuhao-su** [comment]
> > I suggest we preserve a way, which could be other approaches, to enforce the cache size for a period in the near future:
> > 
> > 1. to understand how much better the new dynamic way of adjusting cache size is
> > 2. to get the performance numbers under certain configurations, for demonstration purposes, in case the new way is not behaving as expected
> > 
> > When it gets stabilized, we remove the cache size configuration.
> 
> I would +1 for the idea, but these `SIZE`s are defined with the number of items instead of memory usage i.e. bytes. If you hope to expose them to users, I would recomment to refactor them to limit memory usage instead.
> 
> Otherwise, it seems pretty tricky both for us and users. For example, now we have limited the size of hash map in HashJoin to 65536 but the memory can explode because the rows inside of a join entry was not limited.

+1 for this, the main purpose for this is to limit the memory instead of performance.

**Comment by lmatz** [comment]
>  If you hope to expose them to users, I would recomment to refactor them to limit memory usage instead.

I hope to expose them to developers who want to understand/tune the performance. It is being put in the subsection `[streaming.developer]`.
But I can still do the refactoring for true SIZE in bytes.

> +1 for this, the main purpose for this is to limit the memory instead of performance.

I understand the goal of introducing the new dynamic approach. I just wish I can still tune for performance, i.e. some good numbers to show, in the near future without worrying about the potential complexities/overhead introduced by the new dynamic approach at the beginning.

File: src/common/src/config.rs [hunk 1]
```diff
@@ -360,6 +372,18 @@ mod default {
     pub fn developer_enable_executor_row_count() -> bool {
         false
     }
+
+    pub fn developer_hash_agg_cache_size() -> usize {
```
Comment (fuyufjh) [inline src/common/src/config.rs:376]: How about `unsafe_`?

**Review by fuyufjh — APPROVED** [review]

File: src/common/src/config.rs (hunk 1 above)
Comment (lmatz) [reply src/common/src/config.rs:376]: Sure, and I will refactor to the true SIZE first.

**Pull request merged**